### PR TITLE
Add support for `sys.monitoring` events.

### LINKED ISSUE #9289: [Python 3.12] [Task] Profiler test failures due to Python 3.12 bytecode changes.
Due to Python 3.12 changes the profiler test: `test_profiler.py::TestProfiler::test_profiler` now fails. The fix for this could not be accommodated in https://github.com/numba/numba/pull/9246 (Python 3.12 support update PR) due to technical constraints. 

This issue now tracks the respective test failure and discussions pertaining to it's possible solution.

### PR DESCRIPTION
**Pull request by stuartarchibald**
Python 3.12 introduced a new event monitoring system `sys.monitoring`. This patch augments Numba's dispatcher so as to emit events associated with a Python function starting and returning from execution. Tools monitoring for such events will therefore be able to identify execution of Numba's pure machine code regions as though they were actual Python functions. This allows tools such a ``cProfile`` to record Numba compiled function execution and report them as part of their output.

Fixes #9289

### CODE REVIEW

**Review by sklam — COMMENTED** [review]
I have 

- read through the code to understand the implementation
- manual run cProfile to compare py3.9 vs py3.12 results

Still have to review the code and cross check Python impl.

**Comment by sklam** [comment]
Valgrind showing these when running `python runtests.py numba/tests/test_profiler.py`:

```
==3314481== 8 bytes in 1 blocks are still reachable in loss record 47 of 4,297
==3314481==    at 0x483DF0F: operator new(unsigned long) (vg_replace_malloc.c:483)
==3314481==    by 0x576333F4: Dispatcher_Insert(Dispatcher*, _object*, _object*) (in /home/siu/dev/numba/numba/_dispatcher.cpython-312-x86_64-linux-gnu.so)
==3314481==    by 0x556BF9: method_vectorcall_VARARGS_KEYWORDS (descrobject.c:365)
==3314481==    by 0x546CF0: UnknownInlinedFun (pycore_call.h:92)
==3314481==    by 0x546CF0: PyObject_Vectorcall (call.c:325)
==3314481==    by 0x52D15B: _PyEval_EvalFrameDefault (bytecodes.c:2706)
==3314481==    by 0x581CF9: UnknownInlinedFun (pycore_ceval.h:89)
==3314481==    by 0x581CF9: UnknownInlinedFun (ceval.c:1683)
==3314481==    by 0x581CF9: UnknownInlinedFun (call.c:419)
==3314481==    by 0x581CF9: UnknownInlinedFun (pycore_call.h:92)
==3314481==    by 0x581CF9: method_vectorcall (classobject.c:91)
==3314481==    by 0x53179E: UnknownInlinedFun (call.c:387)
==3314481==    by 0x53179E: _PyEval_EvalFrameDefault (bytecodes.c:3254)
==3314481==    by 0x524E56: UnknownInlinedFun (pycore_ceval.h:89)
==3314481==    by 0x524E56: UnknownInlinedFun (ceval.c:1683)
==3314481==    by 0x524E56: UnknownInlinedFun (call.c:419)
==3314481==    by 0x524E56: _PyObject_FastCallDictTstate (call.c:133)
==3314481==    by 0x562BD5: _PyObject_Call_Prepend (call.c:508)
==3314481==    by 0x63F815: slot_tp_call (typeobject.c:8769)
==3314481==    by 0x520D4A: _PyObject_MakeTpCall (call.c:240)
==3314481==    by 0x52D15B: _PyEval_EvalFrameDefault (bytecodes.c:2706)


==3314481== 3,072 bytes in 1 blocks are still reachable in loss record 4,171 of 4,297
==3314481==    at 0x4844B7C: realloc (vg_replace_malloc.c:1690)
==3314481==    by 0x5070B0: UnknownInlinedFun (obmalloc.c:680)
==3314481==    by 0x5070B0: UnknownInlinedFun (obmalloc.c:1175)
==3314481==    by 0x5070B0: allocate_from_new_pool (obmalloc.c:1389)
==3314481==    by 0x51FF0E: UnknownInlinedFun (obmalloc.c:1553)
==3314481==    by 0x51FF0E: UnknownInlinedFun (obmalloc.c:1564)
==3314481==    by 0x51FF0E: UnknownInlinedFun (obmalloc.c:801)
==3314481==    by 0x51FF0E: UnknownInlinedFun (object.c:332)
==3314481==    by 0x51FF0E: _PyCode_New (codeobject.c:582)
==3314481==    by 0x52820E: r_object (marshal.c:1454)
==3314481==    by 0x527D57: r_object (marshal.c:1223)
==3314481==    by 0x52802F: r_object (marshal.c:1394)
==3314481==    by 0x5F8583: read_object (marshal.c:1525)
==3314481==    by 0x620901: UnknownInlinedFun (marshal.c:1832)
==3314481==    by 0x620901: marshal_loads (marshal.c.h:154)
==3314481==    by 0x52D6DC: _PyEval_EvalFrameDefault (bytecodes.c:2904)
==3314481==    by 0x552783: UnknownInlinedFun (pycore_call.h:92)
==3314481==    by 0x552783: object_vacall (call.c:850)
==3314481==    by 0x581690: PyObject_CallMethodObjArgs (call.c:911)
==3314481==    by 0x580CAC: UnknownInlinedFun (import.c:2779)
==3314481==    by 0x580CAC: PyImport_ImportModuleLevelObject (import.c:2862)
==3314481== 
```

File: numba/_dispatcher.cpp [hunk 1]
```diff
@@ -746,6 +759,290 @@ call_cfunc(Dispatcher *self, PyObject *cfunc, PyObject *args, PyObject *kws, PyO
...
+
+static inline int msb(uint8_t bits) {
+    if (bits > 15) {
+        return MOST_SIG_BIT[bits>>4]+4;
+    }
+    return MOST_SIG_BIT[bits];
+}
+
+
+static int invoke_monitoring(PyThreadState * tstate, int event, Dispatcher *self, PyObject* retval)
```
Comment (sklam) [inline numba/_dispatcher.cpp:784]: I've read this line by line and pondered the refcount. Looks ok.

File: numba/_dispatcher.cpp [hunk 2]
```diff
@@ -746,6 +759,290 @@ call_cfunc(Dispatcher *self, PyObject *cfunc, PyObject *args, PyObject *kws, PyO
...
+    assert(PyCFunction_Check(cfunc));
+    assert(PyCFunction_GET_FLAGS(cfunc) == (METH_VARARGS | METH_KEYWORDS));
+    fn = (PyCFunctionWithKeywords) PyCFunction_GET_FUNCTION(cfunc);
+    tstate = PyThreadState_GET();
+    // issue PY_START if event is set
+    if(invoke_monitoring_PY_START(tstate, self) != 0){
+        return NULL;
+    }
+    // make call
+    pyresult = fn(PyCFunction_GET_SELF(cfunc), args, kws);
```
Comment (sklam) [inline numba/_dispatcher.cpp:1028]:
```suggestion
    pyresult = fn(fn, args, kws);
```

File: numba/_dispatcher.cpp [hunk 3]
```diff
@@ -746,6 +759,290 @@ call_cfunc(Dispatcher *self, PyObject *cfunc, PyObject *args, PyObject *kws, PyO
...
+    if (pyresult == NULL){
+        // TODO: could call monitoring RAISE callbacks here? Would have to know
+        // if the exception is a StopIteration though as that goes via a
+        // different path.
+        return NULL; // exception in Numba call, start to unwind
```
Comment (sklam) [inline numba/_dispatcher.cpp:1033]: I think we have to implement `RAISE`. Consider:

```python
# filename: profile_raise.py
import cProfile as profiler
import os
import pstats

import numpy as np

from numba import jit, types

class MyError(Exception):
    pass


def dot(a, b):
    sum = 0
    for i in range(len(a)):
        sum += a[i]*b[i]
    raise MyError
    return sum


pyfunc = dot


a = np.arange(16, dtype=np.float32)
b = np.arange(16, dtype=np.float32)
cfunc = jit((types.float32[::1], types.float32[::1],))(pyfunc)
# cfunc = pyfunc
n_calls = 123

p = profiler.Profile()
p.enable()
try:
    for _ in range(n_calls):
        try:
            cfunc(a, b)
        except MyError:
            pass
finally:
    p.disable()
stats = pstats.Stats(p).strip_dirs()
code = pyfunc.__code__
expected_key = (os.path.basename(code.co_filename),
                code.co_firstlineno,
                code.co_name,
                )
# check the cfunc (dot) call is in the stats.
assert expected_key in stats.stats
# check that the cfunc (dot) call has been made `n_calls` times.
cfunc_stats = stats.stats[expected_key]
print(cfunc_stats)

# !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
# The error occurs here
# JIT prints: (1, 123, 0.000438581, 0.0005870840000000001, {('profile_raise.py', 13, 'dot'): (122, 1, 0.00034208000000000003, 0.00037650000000000004)})
# PY  prints: (123, 123, 0.001159585, 0.0011721190000000001, {})
assert cfunc_stats[:2] == (n_calls, n_calls)
```

File: numba/tests/test_sys_monitoring.py [hunk 4]
```diff
@@ -0,0 +1,519 @@
...
+        for i in range(6): # there are 5 tool IDs
+            if sys.monitoring.get_tool(i) is not None:
+                monitor_kinds.append(TOOL2MONITORTYPE[i])
+
+        if monitor_kinds:
+            msg = ("Cannot run monitoring tests when other monitors are "
+                   "active, found monitor(s) of type: "
+                   f"{', '.join(monitor_kinds)}")
+            self.skipTest(msg)
+
```
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/tests/test_sys_monitoring.py:63]: RE the `pdb` issue, this problem does reproduce locally, however, it also seems to reproduce if:
1. a test that is added to a different testing module
2. the test in 1. raises an exception
3. the steps above are followed but the new test is run instead.

I'm not sure that this is a problem directly associated with this patch.

File: numba/_dispatcher.cpp (hunk 1 above)
Comment (stuartarchibald) [reply numba/_dispatcher.cpp:784]: Thanks for checking.

File: numba/_dispatcher.cpp (hunk 3 above)
Comment (stuartarchibald) [reply numba/_dispatcher.cpp]: This should be fixed in 4f870eb. It turns out that `cProfile` is listening for a `PY_UNWIND` event associated with a function returning due to an exception being raised.

File: numba/_dispatcher.cpp (hunk 2 above)
Comment (stuartarchibald) [reply numba/_dispatcher.cpp:1061]: I'm not sure that this is correct?

**Comment by sklam** [comment]
BFID `numba_smoketest_cpu_yaml_193`

File: numba/_dispatcher.cpp (hunk 2 above)
Comment (sklam) [reply numba/_dispatcher.cpp:1061]: nvm, i probably misread

**Comment by sklam** [comment]
There's an elusive bug that is only revealed by minimal test sequence:

```bash
python runtests.py numba.tests.test_sys_monitoring.TestMonitoring.test_disable_from_callback numba.tests.test_sys_monitoring.TestMonitoring.test_start_event
```

File: numba/tests/test_sys_monitoring.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (sklam) [reply numba/tests/test_sys_monitoring.py:63]: ok to ignore the pdb problem for now

**Review by sklam — COMMENTED** [review]
https://github.com/numba/numba/pull/9482#issuecomment-2075602944 is the only pending problem and I have no idea why it is happening.

Code and documentation looks great

**Comment by gmarkall** [comment]
> There's an elusive bug that is only revealed by minimal test sequence

I can reproduce this locally too.

**Comment by stuartarchibald** [comment]
Thanks for finding a MWR and helping debug this @sklam, commit 5d652e7 works around the problem.

**Comment by stuartarchibald** [comment]
Note for the future. Were this feature request implemented https://github.com/python/cpython/issues/111997 the code in this patch could be adapted to use it opposed to having to rely on CPython internal details.

File: numba/tests/test_sys_monitoring.py [hunk 5]
```diff
@@ -285,6 +283,14 @@ def test_disable_from_callback(self):
             self.assertEqual(result, self.call_foo_result)
             callback.assert_called()
         finally:
+            # It is necessary to restart events that have been disabled. The
+            # "disabled" state of the `PY_START` event for the tool
+            # `self.tool_id` "leaks" into subsequent tests. These tests then end
+            # up failing as events that should trigger do not! It's not really
```
Comment (sklam) [inline numba/tests/test_sys_monitoring.py:289]:
```
            # up failing as events that should trigger do not! It's not really
                                                      ^^^^^^^
```
you mean `that should not trigger`?

**Review by sklam — COMMENTED** [review]
Workaround confirmed. One minor issue with the code comments

File: numba/tests/test_sys_monitoring.py (hunk 5 above)
Comment (stuartarchibald) [reply numba/tests/test_sys_monitoring.py:289]: The comment is trying to convey that "subsequent tests fail as events in them that should be triggered are not being triggered" (this was why the `Mock.call_count` value was 1 too small, an event that should have triggered did not).

Comment (sklam) [reply numba/tests/test_sys_monitoring.py:289]: My english parser segfaulted when reading it.

**Comment by sklam** [comment]
Reported issue to https://github.com/python/cpython/issues/118327
5d652e70b04a403f6bcc1dd3e349386d9e0e64d7 is the workaround.

File: numba/tests/test_sys_monitoring.py (hunk 5 above)
Comment (stuartarchibald) [reply numba/tests/test_sys_monitoring.py]: I've tried to make this comment more clear in 25535b3. Please take a look and see what you think. Thanks!

**Review by sklam — APPROVED** [review]

**Pull request merged**